src/main/scala/org/constellation/p2p/DataResolver.scala

```diff
@@ -308,107 +181,104 @@ class DataResolver(dao: DAO) {
-      response: List[T] = List.empty[T]
-    ): IO[List[T]] = {
+      response: List[A] = List.empty[A]
+    ): F[List[A]] = {
 
       if (peers.isEmpty && innerHashes.size != allHashes.size) {
-        return IO.raiseError(DataResolutionOutOfPeers(dao.id.short, hashes, sortedPeers.map(_.id.short)))
+        return F.raiseError(DataResolutionOutOfPeers(hashes, sortedPeers.map(_.id.short)))
       }
 
       if (innerHashes.size == allHashes.size) {
-        return response.pure[IO]
+        return response.pure[F]
       }
 
       val peer = peers.head
       val filteredPeers = peers.filterNot(_ == peer)
       val requestHashes = allHashes.filterNot(innerHashes.toSet)
 
-      getBatchData[T](requestHashes, endpoint, peer)(contextToReturn).flatMap { data =>
+      getBatchData[A](requestHashes, endpoint, peer).flatMap { data =>
         logger.debug(
-          s"Requested : ${peer.client.host} : for hashes = ${requestHashes
-            .mkString(",")} : and response hashes = ${data.size}"
+          s"Requested : ${peer.host} : for hashes = ${requestHashes.mkString(",")} : and response hashes = ${data.size}"
         ) >>
-          makeAttempt(filteredPeers, allHashes, innerHashes ++ data.map(_._1), response ++ data.map(_._2))
+          CS.shift >> makeAttempt(filteredPeers, allHashes, innerHashes ++ data.map(_._1), response ++ data.map(_._2))
       }
     }
 
     makeAttempt(sortedPeers, hashes).handleErrorWith(
-      e => logger.error(s"Unexpected error while resolving hashes : ${e.getMessage}") *> IO.raiseError(e)
+      e => logger.error(s"Unexpected error while resolving hashes : ${e.getMessage}") *> F.raiseError[List[A]](e)
     )
   }
 
   private[p2p] def getData[T <: AnyRef](
     hash: String,
-    endpoint: String => PeerResponse[IO, Option[T]],
-    peerApiClient: PeerApiClient
-  )(contextToReturn: ContextShift[IO])(implicit apiTimeout: Duration, m: Manifest[T], dao: DAO): IO[Option[T]] =
-    endpoint(hash).run(peerApiClient.client).onError {
+    endpoint: String => PeerResponse[F, Option[T]],
+    peerApiClient: PeerClientMetadata
+  ): F[Option[T]] =
+    endpoint(hash).run(peerApiClient).onError {
       case _: SocketTimeoutException =>
-        dao.observationService
-          .put(Observation.create(peerApiClient.id, RequestTimeoutOnResolving(List(hash)))(dao.keyPair))
+        observationService
+          .put(Observation.create(peerApiClient.id, RequestTimeoutOnResolving(List(hash)))(keyPair))
           .void
     }
 
-  private[p2p] def getBatchData[T <: AnyRef](
+  private[p2p] def getBatchData[A <: AnyRef](
     hashes: List[String],
-    endpoint: List[String] => PeerResponse[IO, List[(String, Option[T])]],
-    peerApiClient: PeerApiClient
-  )(contextToReturn: ContextShift[IO])(implicit apiTimeout: Duration, m: Manifest[T], dao: DAO): IO[List[(String, T)]] =
+    endpoint: List[String] => PeerResponse[F, List[(String, Option[A])]],
+    peerApiClient: PeerClientMetadata
```

Comment:
You could refactor the name also after changing the type: peerApiClient -> peerClientMetadata